### PR TITLE
[1.8.9] Added CriticalHitEvent

### PR DESCRIPTION
**Pull request by mcenderdragon**
This adds an event to cancel critical player hits and to modify the bonus damage dealt.

### CODE REVIEW

**Comment by mcenderdragon** [comment]
This is for 1.8.9 an closes #2878

**Comment by TehNut** [comment]
This probably won't get pulled into 1.8.9. Retarget at master or 1.9.4.

**Comment by mcenderdragon** [comment]
Well I need it in 1.8.9. And master and 1.9.4 are both a 1.9 version. Can I make a second pull request for these versions ?

**Comment by TehNut** [comment]
1.8.9 is no longer being developed. If you want this in at all, it must be 1.9

**Comment by mcenderdragon** [comment]
Ok then should I create a new pull request for 1.9 ?

File: patches/minecraft/net/minecraft/entity/player/EntityPlayer.java.patch [hunk 1]
```diff
@@ -287,7 +287,24 @@
...
+@@ -1163,9 +1249,15 @@
+                 {
+                     boolean flag = this.field_70143_R > 0.0F && !this.field_70122_E && !this.func_70617_f_() && !this.func_70090_H() && !this.func_70644_a(Potion.field_76440_q) && this.field_70154_o == null && p_71059_1_ instanceof EntityLivingBase;
+ 
++                    net.minecraftforge.event.entity.player.CriticalHitEvent hitResult = null;
++                    if(flag)
++                    {
++                        hitResult = new net.minecraftforge.event.entity.player.CriticalHitEvent(this, (EntityLivingBase)p_71059_1_);
++                        flag = net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(hitResult);
++                    }
```
Comment (williewillus) [inline patches/minecraft/net/minecraft/entity/player/EntityPlayer.java.patch]: this needs to be inverted.
`post` returns TRUE if the event was canceled

File: src/main/java/net/minecraftforge/event/entity/player/CriticalHitEvent.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+@Cancelable
+public class CriticalHitEvent extends AttackEntityEvent
+{
+    public final EntityLivingBase livingTarget;
+    /**
+     * This is how much damage you will deal. By default you deal 150% of you normal damage.
+     * If you set it to 0, then the particles are still generated but you don't do damage.
+    */
+    public float damageModifier;
+    public CriticalHitEvent(EntityPlayer player, EntityLivingBase target)
```
Comment (williewillus) [inline src/main/java/net/minecraftforge/event/entity/player/CriticalHitEvent.java]: no, use private fields and getters/setters
Move doc to the setter

File: patches/minecraft/net/minecraft/entity/player/EntityPlayer.java.patch (hunk 1 above)
Comment (mcenderdragon) [inline patches/minecraft/net/minecraft/entity/player/EntityPlayer.java.patch]: Oh ok, I will chnage it.

**Comment by mcenderdragon** [comment]
Ok i have changed these things, any thing else to do?

**Comment by mcenderdragon** [comment]
I have opend a new pull reqest for 1.9. It is really bad I wasnt fast enough to bring this in the 1.8.9 build but as long as it will be added its fine.

**Comment by Actuarius** [comment]
@williewillus added labels [Feature]

**Pull request closed without merging**